src/AlarmClock.h

```diff
@@ -0,0 +1,140 @@
+   }
+   
+   virtual ~AlarmClock() {
+      mKeepRunning = false;
+      mExited.wait();
+   }
+   
+   bool Expired() {
+      return mExpired.load();
+   }
+
+   int SleepTimeUs() {
+      return kSleepTimeUs;
+   }
+   
+   int SleepTimeMs() {
+      return kSleepTimeMs;
+   }
+
+protected:
+
+   void AlarmClockThread(std::shared_ptr<bool> keepRunning) {
+      SleepTimeIsBelow500ms() ? AlarmClock::SleepForFullAmount() : AlarmClock::SleepInIntervals();
+   }
+   
+   bool SleepTimeIsBelow500ms() {
+      return Duration(kSleepTime) <= milliseconds(kSmallestIntervalInMS);
+   }
+
+   void SleepForFullAmount() {
+      Sleep(kSleepTime);
+      mExpired.store(true);
+   }
+
+   void Sleep(unsigned int sleepTime) {
+      std::this_thread::sleep_for(Duration(sleepTime)); 
+   }
+
+   void Sleep(microseconds t) {
+      std::this_thread::sleep_for(t);
+   }
+
+   void Sleep(milliseconds t) {
+      std::this_thread::sleep_for(t); 
+   }
+   
+   void Sleep(seconds t) {
+      std::this_thread::sleep_for(t); 
+   }
+
+   // If 500ms is NOT an even divisor of the amount of sleep
+   //    time given, should sleep for every multiple of 500ms
+   //    that divides, then check the remaining time and just
+   //    sleep for only that last bit.
+   //
+   // If 500ms IS an even divisor of the sleep time, sleep for 
+   //    all but one multiple of 500ms. There is overhead induced
+   //    by the while loop, so if the full amount of time is slept,
+   //    the precision of the alarm clock will suffer.
+   size_t GetNumberOfSleepIntervals() {
+      return (kSleepTimeMs % kSmallestIntervalInMS == 0) ?
+             ((kSleepTimeMs/kSmallestIntervalInMS) - 1) :
+             (kSleepTimeMs/kSmallestIntervalInMS);
+   }
+  
+   void SleepForRemainder(const unsigned int& currentSleptFor) {
+      if (currentSleptFor < kSleepTimeUs) {
+         Sleep(microseconds(kSleepTimeUs - currentSleptFor));
+      }
+   } 
+
+   void SleepInIntervals() {
+      StopWatch timer;
+      size_t numberOfSleeps = GetNumberOfSleepIntervals();
+      while (KeepRunning() && numberOfSleeps > 0) {
+         Sleep(milliseconds(kSmallestIntervalInMS));
+         --numberOfSleeps; 
+      }
+      auto currentSleptFor = timer.ElapsedUs();
+      if (!KeepRunning()) {
+         mExpired.store(true);
+         return;
+      }
+      SleepForRemainder(currentSleptFor);
+      mExpired.store(true);
+   }
+
+   bool KeepRunning() {
+      return !mExpired.load() && mKeepRunning;
+   }
+   
+   unsigned int ConvertToMilliseconds(Duration t) {
+      return std::chrono::duration_cast<milliseconds>(t).count();
+   }
+   
+   unsigned int ConvertToMicroseconds(Duration t) {
+      return std::chrono::duration_cast<microseconds>(t).count();
+   }
+   
+private:
+
+   std::atomic<bool> mExpired;
+   const int kSleepTime;
+   const int kSleepTimeMs;
+   const int kSleepTimeUs;
+   const int kSmallestIntervalInMS = 500;
+   std::shared_ptr<bool> mKeepRunning;
```

Comment:
As far as I can tell, our version of gcc does not support using ```std::atomic_store(shared_ptr*, shared_ptr)```, so that option is ruled out. 

The two remaining options:

1) ```std::shared_ptr<bool>```
2) ```std::shared_ptr<atomic<bool>>```

Option 1 technically induces a race condition, but Kjell and I discussed this and think the overhead of a race condition would not have enough damaging effects to hinder the system. If the boolean is read as ```false``` but is really ```true```, that's okay, because it will be read properly on the next pass (or eventually). In practice, this proved to be true.

Option 2 guarantees that there will be no race condition, but introduces overhead by using atomics. 

I ran each option 100 times and took the average loop count and precision of the timer, and using a ```shared_ptr<bool>`` won out (the tests always passed obviously).